### ISSUE TITLE
[css-view-transitions-1] CSS selector syntax for generated elements and API names

### ISSUE THREAD

**Issue by jakearchibald** [w3c/csswg-drafts#7788]
Folks at TPAC asked for the pseudo-element names to be reduced so we had a bikeshedding session, and here's what we came up with.

@fantasai @mirisuzanne I'm interested to hear your thoughts on this. We've tried to go for brevity rather than full namespacing.

## "Shared element transition" becomes "View transition"

Rationale:

- "shared element" didn't seem accurate. There's no sharing of elements.
- The DOM changes, and a transition is created from the two states. The actual DOM elements aren't animated, it's just a view of them, so "view transitions".

## `page-transition-tag` becomes<br/>`view-name`

```js
.header {
  /* old */
  page-transition-tag: header;
  /* new */
  view-name: header;
}
```

This causes pseudos to be created so the header can be animated independently during the transition.

Rationale:

- The scope of a transition may not always be the whole "page".
- Other specs use "name" rather than "tag" to set an identifier (`container-name`, `scroll-timeline-name`).
- Using "view" because "view transition".

## `html::page-transition` becomes<br/> `html::view-root`

This is the root of the pseudo-element tree that contains all the transition parts.

Rationale:

- The scope of a transition may not always be the whole "page".
- Using "view" because "view transition".
- "root" because it's the root of the pseudo-element tree.

## `html::page-transition-container(header)` becomes<br/> `html::view(header)`

This is the pseudo-element that represents a view of an element during the transition. In this case it's the view of the 'header'. Developers target this to customize the animation. It's a child of `html::view-root`.

Rationale:

- Each element that's given a `view-name` results in the creation of one of these pseudo-elements. So `view-name: header` results in a `html::view(header)` during the transition.

## `html::page-transition-image-wrapper(header)` becomes<br/> `html::view-image-group(header)`

This wraps the view's images, providing isolation for their blending (to allow a true cross-fade). In this case it's the image-group for the 'header'. It's a child of `html::view`.

Rationale:

- The pseudo elements within are replaced elements, but they behave like images. They have a width, height, and an aspect ratio.

## `html::page-transition-outgoing-image(header)` and<br/> `html::page-transition-incoming-image(header)` become<br/> `html::view-before(header)` and<br/> `html::view-after(header)` respectively

These are the replaced elements that represent the before and after view. In their default animation, they cross-fade.

Rationale:

- Folks had difficulty mapping 'outgoing' and 'incoming'.
- 'before' and 'after' seem more intuitive, and shorter.
- The web animation API uses language like "before phase" and "after phase", although this is only in the spec and not exposed in the API.

## `document.createTransition()` becomes<br/> `document.createViewTransition()`

This is the API developers use to create a transition between states in an SPA.

Rationale:

- Although it's longer, it removes the ambiguity with CSS transitions.

## Example

Taking the slide-from-the-side example from the breakout at TPAC:

Old:

```css
::page-transition-outgoing-image(root) {
  animation: 500ms ease-out both slide-to-left;
}

::page-transition-incoming-image(root) {
  animation: 500ms ease-out both slide-from-right;
}
```

New:

```css
::view-before(root) {
  animation: 500ms ease-out both slide-to-left;
}

::view-after(root) {
  animation: 500ms ease-out both slide-from-right;
}
```

(I've omitted the keyframes since they're the same in both)

## Resulting pseudo-tree

```
::view-root
├ ::view(custom-ident)
|  └─ ::view-image-group(custom-ident)
|     ├─ ::view-before(custom-ident)
|     └─ ::view-after(custom-ident)
└ ::view(different-ident)
   └─ ::view-image-group(different-ident)
      ├─ ::view-before(different-ident)
      └─ ::view-after(different-ident)
```

(thanks @mirisuzanne!)

**Comment by bramus** [comment]
Your last example uses `::view-before` twice. Second occurrence should be `::view-after`.

**Comment by jakearchibald** [comment]
Thanks, fixed!

**Comment by ydaniv** [comment]
`View` may be confusing since there's already [`ViewTimeline`](https://drafts.csswg.org/scroll-animations-1/#view-timelines) and its semantics are quite different - closer to those of `viewport` and `viewbox`.
And if considering `view()` it may also be relevant to consider #7587.

I guess a more accurate name would be "Cross Element Transition", but that isn't helping with the brevity part 🙂 
Perhaps something like `cross-transition`? (or `x-transition`?)

**Comment by tbondwilkinson** [comment]
I am quite fond the name "view"!

`view-name` sounds like it's the name of the view, but really it's a "part" of the overall "view" transition. 

Just so I understand, why is there a need to use a new tagging system, rather than reusing one of the existing tagging systems (class, id)? I assume it's hard to support general selectors in psuedoelements here, but still wondering whether with this new DOM structure being represented in pseudoelements, rather you really do need some sort of selector syntax for that. I believe this was proposed elsewhere, I forget what the response was.

```
html::view(.header)
```

I do like the shorter names, but naming things just `::view` does lose the sense of transition. `::view` alone sounds like it could be static, like `::backdrop` rather than something ephemeral and specific to a transition. Somewhere in between is `::view-transition`

Alternate to `before` and `after` is `enter` and `exit`, which are slightly shorter.

Rather than `view-before` is a pseudoclass an option?

```
::view-transition(root):enter
```

**Comment by jakearchibald** [comment]
@tbondwilkinson 

> `view-name` sounds like it's the name of the view, but really it's a "part" of the overall "view" transition.

`view-part-name` may be more accurate, but longer.

> Just so I understand, why is there a need to use a new tagging system, rather than reusing one of the existing tagging systems (class, id)?

I don't think setting a class or an ID on an element means it should be a seperate part in an animated transition. That isn't what those things semantically mean. And with class especially, it would cause an explosion of layers, since it's extremely common for an element to have a class.

It's common to want to differ the animation based on:

- Viewport width
- User preference (eg prefers-reduced-motion)
- Cascade (like a class name on the root element may influence whether a descendant is a seperate part of an animation)

This is trivial with CSS (`@media`, the cascade), but really really had to do with attributes.

Besides, this is very much a 'design' feature, so it semantically fits better with CSS than DOM attributes. See also `container-name` and `scroll-timeline-name`.

> but still wondering whether with this new DOM structure being represented in pseudoelements, rather you really do need some sort of selector syntax for that. I believe this was proposed elsewhere, I forget what the response was.

Yeah, we're working on a few options here. I'll create a separate issue for that. I'm waiting on Mozilla folks to review it, because I want to be sure I'm representing their proposal correctly. That shouldn't impact this naming change though.

> Rather than `view-before` is a pseudoclass an option?

It would work, but I don't see the advantage. This would require `::view-transition(root)` to match:

- the root part
- the root part's image group
- the root part's before image
- the root part's after image

…and I don't see why you'd ever want to do that.

**Comment by mirisuzanne** [comment]
Bikeshedding overall name can keep happening - short is nice, and clear is also nice, and there's probably a balance worth finding. 

I'm also interested in the structure being represented. I know one of the alternatives we discussed at TPAC involved some sort of 'pseudo-element descendant combinators' to make that structure more clear, and I pushed back some on needing a whole new syntax for it. But it's true that the structure gets a bit lost in this variant. 

I think that the resulting pseudo-element tree is:

```
::view-root
├ ::view(custom-ident)
|  └─ ::view-image-group(custom-ident)
|     ├─ ::view-before(custom-ident)
|     └─ ::view-after(custom-ident)
└ ::view(different-ident)
   └─ ::view-image-group(different-ident)
      ├─ ::view-before(different-ident)
      └─ ::view-after(different-ident)
```

My immediate instinct was to try and put some structure inside the functional notation somehow:

```
::view-root
├ ::view(custom-ident)
|  └─ ::view(custom-ident image-group)
|     ├─ ::view-image(custom-ident image-before)
|     └─ ::view-image(custom-ident image-after)
...
```

But looking at it, I'm not convinced that's a good path to take. Maybe the real gain I'm looking for here is just to associate the 'image group' with the 'before' and 'after' images a bit more clearly. Something like:

```
::view-root
├ ::view(custom-ident)
|  └─ ::view-image-group(custom-ident)
|     ├─ ::view-image-start(custom-ident)
|     └─ ::view-image-end(custom-ident)
...
```

or:

```
::view-root
├ ::view(custom-ident)
|  └─ ::view-group(custom-ident)
|     ├─ ::view-start(custom-ident)
|     └─ ::view-end(custom-ident)
...
```

So in the end, I like the direction you're going - and it maybe does just come down to a bit more bikeshedding. I kinda like `start` and `end`, since we already have `before` and `after` pseudos that mean something pretty different. But I don't feel strongly.

**Comment by tbondwilkinson** [comment]
> Just so I understand, why is there a need to use a new tagging system, rather than reusing one of the existing tagging systems (class, id)?

Let me rephrase, why not have an opt-in to transitions and then reuse whatever class/id already exists on that element.

```
<div id="foo"></div>
```

```
#foo {
  view-transition: root;
}
```

Which leads to something like:
```
::view-transition(root #foo):enter
```

**Comment by jakearchibald** [comment]
@mirisuzanne 

> I'm also interested in the structure being represented. I know one of the alternatives we discussed at TPAC involved some sort of 'pseudo-element descendant combinators' to make that structure more clear

Yeah, I've got a whole other doc on that with all the options that I'd love your opinion on. I'm waiting on @emilio to review it first, because I want to make sure I've represented the shadow DOM pattern correctly & fairly.

> I think that the resulting pseudo-element tree is:
> 
> ```
> ::view-root
> ├ ::view(custom-ident)
> |  └─ ::view-image-group(custom-ident)
> |     ├─ ::view-before(custom-ident)
> |     └─ ::view-after(custom-ident)
> └ ::view(different-ident)
>    └─ ::view-image-group(different-ident)
>       ├─ ::view-before(different-ident)
>       └─ ::view-after(different-ident)
> ```

Yep, that's correct

> I kinda like `start` and `end`, since we already have `before` and `after` pseudos that mean something pretty different. But I don't feel strongly.

Hmm, `start` and `end` feel layout-positional to me (from `align-content` and co). I get what you mean about `::before` and `::after` though.

**Comment by mirisuzanne** [comment]
> Hmm, `start` and `end` feel layout-positional to me (from `align-content` and co). I get what you mean about `::before` and `::after` though.

Yeah, I agree that both have baggage. In my mind the `start`/`end` baggage is just a bit more _distant_, since it's not part of existing pseudo-classes? But it's maybe a small distinction.

**Comment by jakearchibald** [comment]
Both are better than `incoming` and `outgoing` anyway

**Comment by fantasai** [comment]
Hi @jakearchibald! I love the thought that you put into this. It's maybe obvious, but I'm missing why ::view(header) and ::view-image-group(header) are two different elements? Can we combine them?

I agree with @SebastianZ that we seem to have lost the transition aspect of the names, and maybe that's worth keeping somehow.

If you want to use pseudo-element structure to represent the nesting, maybe something like:

```
::transition /* root */
::transition::view(header) /* group */
::transition::view(header)::old   /* outgoing image */
::transition::view(header)::new /* incoming  image */
```

If not, then collapsing down the names:
```
::transition
::transition-view(header)
::transition-view-old(header)
::transition-view-new(header)
```

**Comment by astearns** [comment]
`transition` is long and the feature actually uses animations. `animation` is also long. `view` by itself doesn’t say what is being done to the view. What about `view-swap` which is short, easy to spell, and says what is being done (but is vague on the details of how)

**Comment by fantasai** [comment]
Just throwing out some ideas...

Using "swap" instead of "transition" as @astearns suggested and adding an image-pair wrapper to support nesting going forward...

```
::swap
::swap-view(header)
::swap-view-group(header)
::swap-view-old(header)
::swap-view-new(header)
```

or maybe

```
::view-swap
::view-swap-set(header)
::view-swap-pair(header)
::view-swap-old(header)
::view-swap-new(header)
```

(I don't quite like the confusion over whether sets or pairs are outermost.)

I think View Transitions is a very natural name for this feature, though, and keeping it in the pseudo names is long but also helpful?

**Comment by khushalsagar** [comment]
> I don't quite like the confusion over whether groups or sets are outermost.

Would `::swap-view-images` or `::view-swap-images` make that clearer? Since the only nodes underneath it would be the old and new images. The flip side is, it can be odd for a single node to have plural images in the name.

> I think View Transitions is a very natural name for this feature, though, and keeping it in the pseudo names is long but also helpful?

Between transition and swap, I like transition better too. It's not using CSS transitions but it's conceptually doing something similar.

**Comment by jakearchibald** [comment]
(on holiday, so sorry for the brief comment)

Yeah, "swap" to me sounds like the opposite of a transition. As "the visual state just swaps, rather than transitions".

**Comment by jakearchibald** [comment]
@fantasai 

> I think View Transitions is a very natural name for this feature, though, and keeping it in the pseudo names is long but also helpful?

Yeah, I'm thinking:

```
view-transition-name:

::view-transition-root
::view-transition(name)
::view-transition-images(name)
::view-transition-old(name)
::view-transition-new(name)

document.createViewTransition();
```

It doesn't do much for brevity, but I think it's pretty descriptive.

I've gone with `::view-transition(header)` for brevity, since it'll be more common to select that than the root, which I've given `::view-transition-root`.

**Comment by fantasai** [comment]
@jakearchibald That looks pretty self-consistent! A couple minor things I'd modify:

- We cold drop -root, and just have `::view-transition` without parens be the root
- I think I like `::view-transition-set` better than `::view-transition-images`

So

```
view-transition-name:

::view-transition
::view-transition(name)
::view-transition-set(name)
::view-transition-old(name)
::view-transition-new(name)

document.createViewTransition();
```

How well does this structure work with your nesting proposal?

**Comment by khushalsagar** [comment]
> We cold drop -root, and just have ::view-transition without parens be the root

That makes things ambiguous. Especially because we do want the function syntax to allow selecting all pseudo-elements for a particular type. For example:

```css
::view-transition
::view-transition(*)
::view-transition-set(*)
::view-transition-old(*)
::view-transition-new(*)
```

The fact that `::view-transition(*)` and `::view-transition` selects different elements is not intuitive. `::view-transition-root` is longer but better in this regard.

> document.createViewTransition();

See related discussion in #7828, is document.transition() or document.transitionView() better to make it obvious that it also starts the transition (like element.animate()).

> How well does this structure work with your nesting proposal?

Nesting could allow us to drop the transition word since it would be obvious in the syntax:

```css
::view-transition-root
::view-transition-root::view(name)
::view-transition-root::view(name)::view-set
::view-transition-root::view(name)::view-set::view-old
::view-transition-root::view(name)::view-set::view-new
```

But the css function syntax in the first one seems much simpler to directly target these elements. Do you think it's better to just go with those? If we do resolve on backing these with a pseudo-element tree (instead of shadow DOM) then the nested syntax will work naturally (similar to other pseudo-elements) but with the longer name. I think that's fine since developers will likely use the function syntax.

```css
::view-transition-root::view-transition(name)::view-transition-set::view-transition-old
```

**Comment by astearns** [comment]
> Yeah, "swap" to me sounds like the opposite of a transition. As "the visual state just swaps, rather than transitions".

That’s fair. I am just throwing out shorter, vaguer options to test out the theory that `view-transition` is the best we can do. What about `view-change`?

**Comment by fantasai** [comment]
> Nesting could allow us to drop the transition word since it would be obvious in the syntax:

I was referring to the proposal described in https://github.com/w3c/csswg-drafts/issues/7788#issuecomment-1261074637 not actually structuring the pseudo-elements.

If we're going to use pseudo-element structure, we should drop a lot more of the redundancies in the pseudo-element names, since they'll be scoped by their parent.

E.g.
```
::transition
::transition::view(name)
::transition::view(name)::set
::transition::view(name)::set::old
::transition::view(name)::set::new
```

The main concern I have is how this all plays with the nested transitions extension.

**Comment by jakearchibald** [comment]
I like `::view-transition-images` because the two things in it behave like images. They have a width, height, and aspect ratio. `inset: 0` behaves the same way it would on an image.

However, this element, and the root, are the ones that I expect to be selected the least. So I'm happy with longer more descriptive names there.

Really appreciate all the effort bikeshedding this! I'll dig in more next week.

**Comment by vmpstr** [comment]
If I may summarize, there didn't seem to be an objection to generally having `view-transition-` prefix for the pseudo elements, but admittedly we didn't focus the discussion on this.

Most of the discussion centered around `view-transition-images` in [this proposal](https://github.com/w3c/csswg-drafts/issues/7788#issuecomment-1266772511)

* There was some pushback on using the word "image" or "images" in here, since the element is not an image but a container. The counter proposal was `view-transition-set`
* The concern about "set" was that it's too generic

Some other proposals I saw in no particular order were:
* `view-transition-images`
* `view-transition-set`
* `view-transition-image-set`
* `view-transition-effect-group`

and some other combinations of those words.

In this domain, we don't expect this element to be frequently targeted, so I think a longer name is acceptable.

Personally, I'm fine with one of the four combinations: `view-transition-{image,effect}-{group,set}`, which seems to be most of the proposals anyway :)

Do folks have other front runners that they would like to consider? Or is there a combination of those that is particularly objectionable? I'm hoping we can reduce the set to a handful of candidates and take a vote if there is no consensus otherwise.

**Comment by khushalsagar** [comment]
Reading the notes from the discussion, I think `view-transition-effect-group` would address concerns with the other options. It's not using the word image which creates confusion about the elements underneath being replaced elements but not `img`. And avoids the too generic concern with just view-transition-group. It's longer but between brevity and clarity, we should err on the side of brevity.

@fantasai @astearns @tabatkins, any objection to `view-transition-effect-group`?

**Comment by tabatkins** [comment]
I continue to object to the idea that calling the images "images" is somehow confusing. The child elements *are* images, both literally and semantically; the image-focused object-* properties apply to them. I still don't understand what @fantasai was trying to say in her argument against it.

Calling it **image** would indeed be confusing and wrong, since it's a container for the images, not an image itself. But "images" is fine; the plural clearly indicates it's a container, and its children are the images being discussed.

Ultimately, as you say, this element will be rarely targeted, so its name doesn't matter that much, I just don't want us to deliberately choose a bad name for invalid reasons.

**Comment by khushalsagar** [comment]
> The child elements are images, both literally and semantically; the image-focused object-* properties apply to them.

Is literally correct here? The content for these replaced elements can be a vector graphic which the browser can re-raster if the scale changes. That aligns with SVG or 2D canvas. object-* properties apply to these elements as well.

**Comment by tabatkins** [comment]
A vector graphic is an image, yes. ^_^

**Comment by vmpstr** [comment]
> I continue to object to the idea that calling the images "images" is somehow confusing.

The plurality of the name is confusing to me. This is referring to a single element, which indeed is a container of images, but it's a container (singular), it's not images. When I'm selecting this element, and I have `view-transition-images(...) { ... }` I know that this is a single element, because I worked on this feature :) but if I didn't, I don't know if this is some sort of magic CSS to select all constituent images.

To draw a contrived analogy, it's kind of like instead of `<ol>` calling it `<ordereditems>`. It isn't items, it's a list (singular) of items and when selecting `ol { }` i'm selecting a single container element vs `ordereditems { }` selecting seemingly multiple items

Personally, I don't have any objections to using the word "image", but I want this to be something like `image-group` or `image-list`, not `images`

**Comment by vmpstr** [comment]
In the interest of summarizing this for a in-person discussion, below is the list of elements that we need to have a name for:

* view transition root, the top candidate seems to be `::view-transition`
* the container element, which is the child of the root and has size/transform of the underlying element on it. The top contender is `::view-transition-container`
* The wrapper element, which has isolation and contains in it one or two images. This doesn't have consensus, with the most of the discussion being about this element
* the outgoing image from the "old" state. Top choice: `::view-transition-old`
* the incoming image from the "new" state. Top choice: `::view-transition-new`
* the function to initiate the transition, currently `createDocumentTransition`. Since this is also now specced to begin the transition, I prefer a verb name with the proposal: `document.transitionView()` read as a verb, not a noun, which is analogous to `element.animate()`. 

----

Also, we don't have an agreement on the selector syntax
**Option 1** This should be chained like:
`html::view-transition::view-transition-container(foo)::view-transition-some-wrapper(foo)::view-transition-old(foo)`
**Option 2** presumably we can drop "foo" and "view-transition" prefix from nested elements:
`html::view-transition::container(foo)::some-wrapper::old`

**Option 3** or do we select this as if they are all children of html: 
`html::view-transition`, `html::view-transition-old(foo)`, etc

**Option 4** or do we have part-like selector function that just "figures it out":
`::view-transition-part(container foo)`, `::view-transition-part(old foo)`

The last option is also related to #7928 

My proposal is to chain this to do option 1 and option 4. Chain it like this:
`html::view-transition::view-transition-container(foo)::view-transition-some-wrapper(foo)::view-transition-old(foo)`
and also provide a convenience part-like function:
`view-transition-part(container foo)`

**Comment by khushalsagar** [comment]
Thanks for summarizing this Vlad! One thing missing from the list above is the CSS property used to tag a DOM element. I think we have consensus there on `view-transition-name`. 

I'll post a combination of name + selector syntax from the options above for each pseudo-element. So we can have a precise resolution:

### Option 1 (Full chaining)
```css
/* root for all pseudos */
html::view-transition
/* container for DOM elements with tag foo */
html::view-transition::view-transition-container(foo)
/* Isolation node for the 2 replaced elements */
html::view-transition::view-transition-container(foo)::view-transition-image-group(foo)
/* Replaced element for old content */
html::view-transition::view-transition-container(foo)::view-transition-image-group(foo)::view-transition-old(foo)
/* Replaced element for new content */
html::view-transition::view-transition-container(foo)::view-transition-image-group(foo)::view-tranition-new(foo)

/* Each selector with args uses * to select all pseudo-elements of that type */
html::view-transition::view-transition-container(*)
html::view-transition::view-transition-container(*)::view-transition-image-group(*)
html::view-transition::view-transition-container(*)::view-transition-image-group(*)::view-transition-old(*)
html::view-transition::view-transition-container(*)::view-transition-image-group(*)::view-transition-new(*)
```

### Option 2 (Reduced chaining)
```css
/* root for all pseudos */
html::view-transition
/* container for DOM elements with tag foo */
html::view-transition::container(foo)
/* Isolation node for the 2 replaced elements */
html::view-transition::container(foo)::image-group
/* Replaced element for old content */
html::view-transition::container(foo)::image-group::old
/* Replaced element for new content */
html::view-transition::container(foo)::image-group::new

/* Each selector with args uses * to select all pseudo-elements of that type */
html::view-transition::container(*)
html::view-transition::container(*)::image-group
html::view-transition::container(*)::image-group::old
html::view-transition::container(*)::image-group::new
```
This option is terse but we'd end up using `old` and `new` as pseudo-element names which seem too generic.

### Option 3 (Direct html selector)
```css
/* root for all pseudos */
html::view-transition
/* container for DOM elements with tag foo */
html::view-transition-container(foo)
/* Isolation node for the 2 replaced elements */
html::view-transition-image-group(foo)
/* Replaced element for old content */
html::view-transition-old(foo)
/* Replaced element for new content */
html::view-transition-new(foo)

/* Each selector with args uses * to select all pseudo-elements of that type */
html::view-transition-container(*)
html::view-transition-image-group(*)
html::view-transition-old(*)
html::view-transition-new(*)
```

### Option 4 ([Part](https://w3c.github.io/csswg-drafts/css-shadow-parts/#part) like)
```css
/* root for all pseudos */
html::view-transition
/* container for DOM elements with tag foo */
html::view-transition-part(container foo)
/* Isolation node for the 2 replaced elements */
html::view-transition-part(image-group foo)
/* Replaced element for old content */
html::view-transition-part(old foo)
/* Replaced element for new content */
html::view-transition-part(new foo)

/* Specifying the part associated with only the element type selects all elements of that type */
html::view-transition-part(container)
html::view-transition-part(image-group)
html::view-transition-part(old)
html::view-transition-part(new)
```

**Comment by fantasai** [comment]
Possible suggestion: changing `::view-transition-image-group` to `html::view-transition-pair`

**Comment by jakearchibald** [comment]
TODO: capture @fantasai & @mirisuzanne's suggestion here

**Comment by jakearchibald** [comment]
@fantasai and @mirisuzanne suggested a syntax like this:

```css
html::view-transition(.header new) {
  …
}
```

Where the bit within `view-transition()` would allow for a complex selector where the `view-transition-name` is 'turned into' a class name, and the parts of the structure are 'turned into' tag names. So the above is selecting 'new' parts within the container with `view-transition-name` of `header`.

**Comment by FremyCompany** [comment]
Note that if we want to allow complex selectors like this in the future, but not in the first version, this is possible if we change the syntax to allow ID selectors only:
```css
html::view-transition(#image-group-foo) {
  …
}
html::view-transition(#image-group-foo) {
  …
}
```
(where the ids are auto-generated based on the syntax that would be used in proposal 4)

That enable us to go for more complex selectors later, if there is a demand, without adding much difficulty for authors (only add one pound symbol in the function).

Of course, if we are quite sure we won't need complex selectors, this is moot. Just a random thought.

**Comment by khushalsagar** [comment]
Filed #7960 since we didn't get to resolve on all the names today.

@jakearchibald regarding the syntax you mentioned, when you say tag names are you drawing a parallel with element tag like `img`. So the proposal there would be equivalent to a selector like this in CSS:

```css
.class-foo img {
  ...
}
```

I assumed this proposal is about allowing existing selector patterns as the function argument but the above is not valid syntax today. Maybe I misunderstood the proposal.

@FremyCompany that's an interesting thought. ID is not a good concept though since it'll be unique per pseudo-element and we do want to allow selecting all containers for example. Classes is better. But I'm a little wary of supporting all complex selectors there.

The use-cases for more complex selectors we've seen have been about rules which are conditional based on the DOM structure. Could we do that by allowing a subset of existing pseudo-classes on pseudo-elements. So something like `html::view-transition-old(foo):only-child` if the styles are conditional based on there being a pair of images vs an exit transition (only old content).

**Comment by tabatkins** [comment]
Yes, the idea suggested by fantasai/miriam is that we expose the structure of the pseudo tree directly, with the names mapped to element names and the tags mapped to classes. So `html::view-transition(.foo old)` matching an "old" image who is a descendant of a "foo"-tagged container or image-group. Then you could write `html::view-transition(old.foo:only-child)` to get an old that doesn't have a new.

**Comment by mirisuzanne** [comment]
I'll note that our suggestion doesn't rely specifically on the specific mapping of class/element syntax, but that provides a good example. Whatever the selector-syntax mapping (@FremyCompany's suggestion of IDs might make more sense), the goal is just to allow compound/complex selectors in the function.

One reason for that might be exposing the hierarchy. The more important goal in my mind was the ability to differentiate what happens in these three instances:

1. There is _only a `new` image_.
2. There is _only an `old` image_.
3. There are _both `old` and `new` images_.

The solution mentioned a few times for that is using `:only-child`/:not(:only-child)` on the element in question. I assume that solution is also possible with the current proposal as well?

```css
html::view-transition-old(*):only-child {
  /* this is a default exit animation */
}
```

**Comment by jakearchibald** [comment]
@mirisuzanne I think we need to make that work. And if it can't be `:only-child`, we should add something view-transition-specific to do the same thing.

**Comment by khushalsagar** [comment]
I think we should be able to make the :only-child pseudo-class work. I figured pseudo-classes on pseudo-elements would already be a supported concept. For instance, before:hover seems like a valid use-case. But I just tried and it's not supported. Maybe there is an implementation complexity associated with supporting pseudo-classes on pseudo-elements in general that I'm missing.

**Comment by jakearchibald** [comment]
I think it's tricky because what does `::before:only-child` mean? We might need to allow-list which pseudos can use `:only-child`.

**Comment by khushalsagar** [comment]
The spec edits for this issue have landed. I started a different one to capture the use-case discussed above.

In case I missed anything else on this issue which still needs discussion please file another issue for it.
